extensions/amp-selector/1.0/base-element.js

```diff
@@ -0,0 +1,290 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {Option, Selector} from './component';
+import {PreactBaseElement} from '../../../src/preact/base-element';
+import {
+  closestAncestorElementBySelector,
+  createElementWithAttributes,
+  toggleAttribute,
+  tryFocus,
+} from '../../../src/dom';
+import {dev, devAssert, userAssert} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+import {toArray} from '../../../src/types';
+import {useCallback, useLayoutEffect, useRef} from '../../../src/preact';
+
+export class BaseElement extends PreactBaseElement {
+  /** @override */
+  init() {
+    const {element} = this;
+    this.optionState = [];
+
+    // Set up API
+    this.registerApiAction('clear', (api) => {
+      api./*OK*/ clear();
+      this.mutateProps(dict({'value': []}));
+    });
+
+    this.registerApiAction('selectUp', (api, invocation) => {
+      const {args} = invocation;
+      const delta = args && args['delta'] !== undefined ? -args['delta'] : -1;
+      api./*OK*/ selectBy(delta);
+    });
+    this.registerApiAction('selectDown', (api, invocation) => {
+      const {args} = invocation;
+      const delta = args && args['delta'] !== undefined ? args['delta'] : 1;
+      api./*OK*/ selectBy(delta);
+    });
+
+    this.registerApiAction('toggle', (api, invocation) => {
+      const {args} = invocation;
+      const {'index': index, 'value': opt_select} = args;
+      userAssert(typeof index === 'number', "'index' must be specified");
+      const option = this.optionState[index];
+      if (option) {
+        api./*OK */ toggle(option, opt_select);
+      }
+    });
+
+    // Listen for mutations
+    const mu = new MutationObserver(() => {
+      if (this.isExpectedMutation) {
+        this.isExpectedMutation = false;
+        return;
+      }
+      const {children, options} = getOptions(element, mu);
+      this.optionState = options;
+      this.mutateProps({children, options});
+    });
+    mu.observe(element, {
+      attributeFilter: ['option', 'selected', 'disabled'],
+      childList: true,
+      subtree: true,
+    });
+
+    // Return props
+    const {children, value, options} = getOptions(element, mu);
+    this.optionState = options;
+    return dict({
+      'as': SelectorShim,
+      'shimDomElement': element,
+      'children': children,
+      'value': value,
+      'options': options,
+    });
+  }
+}
+
+/**
```

Comment:
Nothing below this line was altered; all cut-pasted directly from `amp-selector.js`